packages/block-library/src/comment-template/index.php

```diff
@@ -5,6 +5,84 @@
  * @package WordPress
  */
 
+/**
+ * Convert the flat array of comments to a nested array
+ * containing comments with a `comment_children` property.
+ *
+ * @param WP_Comment[] $comments  - Array of comments.
+ * @return array
+ */
+function convert_to_tree( $comments ) {
+	$table = array();
+
+	// If there are no comments, we can return early an empty array.
+	if ( ! isset( $comments ) || empty( $comments ) ) {
+		return array();
+	}
+
+	// First we create an associative array of the form:
+	// [commentId] => [...comment, comment_children: [] ].
+	foreach ( $comments as $comment ) {
+		$comment_with_children                     = (array) $comment;
+		$comment_with_children['comment_children'] = array();
+		$table[ $comment->comment_ID ]             = $comment_with_children;
+	}
+
+	$tree = array();
+
+	foreach ( $comments as $comment ) {
+		if ( ! empty( $comment->comment_parent ) ) {
+			// If the comment has a "parent", then find that parent in the table that
+			// we have created above and push the current comment to the array of its
+			// children.
+			$table[ $comment->comment_parent ]['comment_children'][] = $table[ $comment->comment_ID ];
+		} else {
+			// Otherwise, if the comment has no parent (also works if parent is 0/"0")
+			// that means that it's a top-level comment. So we can find it in the table
+			// and push it to the final tree.
+			//
+			// We have to push by reference using `=&` because this top-level comment might
+			// still have other children. This way, if we add another comment to the
+			// children of the current comment (in the `if` branch), it will be
+			// updated in the final tree.
+			$tree[] =& $table[ $comment->comment_ID ];
```

Comment:
I must admit I have a hard time following this trick. I will try to process this logic and see if I can find any alternative.

Reply:
Oh, this function is not necessary anymore because I figured out that I can use the `get_approved comments()` with the `'hierarchical' => 'threaded'` parameter to get all comments in a "tree" format directly from WordPress: 

https://github.com/WordPress/gutenberg/blob/67d8d57f009d1098f86705015b384181541b7106/packages/block-library/src/comment-template/index.php#L67-L73

And then I just need to call `get_children()` on each level of nested comments:

https://github.com/WordPress/gutenberg/blob/67d8d57f009d1098f86705015b384181541b7106/packages/block-library/src/comment-template/index.php#L26

The updates I've pushed implement this, you can take a look 🙂 

By the way, the PHP implementation had exactly the same logic as the JS one, but perhaps there was a more idiomatic way to do this in PHP - I just "translated" it from JS 😅

Reply:
Just noting that this also means that I now have to update the JS side to behave the same way as the `get_approved_comments()` with `'hierarchical' => 'threaded'`. Why?

The `queryPerPage` attribute determines the number of _top-level_ comments that are being shown (which I think is actually the correct way). However, currently on the JS-side `queryPerPage` determines the _total_ number of comments, including nested comments. 

So, if `queryPerPage` is `2` and we have comments like:
```
└─ comment 1
   └─ comment 2
        └─ comment 4
└─ comment 3 
```

In, the admin, only comments 1&2 will be shown, whereas on the frontend all comments will be shown.

Reply:
All right, this was easily handled in 393b2c1